File: Sources/structly/03_beginner_recursion/SumNumbersRecursive.swift
//
//  SumNumbersRecursive.swift
//  MyCLI
//
//  Created by new on 3/14/26.
//

func sumNumbersRecursive(_ numbers: [Int]) -> Int {
    func sum(i: Int, total: Int) -> Int {
        if i == numbers.count {
            return total
        }
        return sum(i: i+1, total: total + numbers[i])
    }
    return sum(i: 0, total: 0)
}

func sumNumbersRecursive1(_ numbers: [Int]) -> Int {
    func sum(_ numbers: ArraySlice<Int>) -> Int {
        if numbers.isEmpty {
            return 0
        }
        return numbers[numbers.startIndex] + sum(numbers.dropFirst())
    }
    return sum(numbers[...])
}

File: Sources/structyC/03_beginner_recursion/sum_numbers_recursive.cpp
//
//  sum_numbers_recursive.cpp
//  MyCLI
//
//  Created by new on 3/14/26.
//

#include <vector>

int sumNumbersRecursive(std::vector<int> numbers) {
    if (numbers.size() == 0) {
        return 0;
    }
    return numbers[0] + sumNumbersRecursive(std::vector<int>(numbers.begin()+1, numbers.end()));
}

int sum(int i, int total, std::vector<int> numbers) {
    if (i == numbers.size()) {
        return total;
    }
    return sum(i+1, total + numbers[i], numbers);
}

int sumNumbersRecursive1(std::vector<int> numbers) {
    return sum(0, 0, numbers);
}

